File: evd_unity_sim_app/Assets/Application/_Scripts/EvD Script Model/data/Location.cs
using System.Collections;
using System.Collections.Generic;


namespace EvD
{
    namespace Data
    {

        [System.Serializable]
        public class Location : Waypoint
        {

            /*
            * Constructors
            */
            
            public static new string TypeString()
            {
                return "location.";
            }

            public static new string FullTypeString()
            {
                return Waypoint.FullTypeString() + TypeString();
            }
            
            public Location(Position position = null, Orientation orientation = null, float[] joints = null, 
                            string type="", string name = "", string uuid = null, Node parent = null, 
                            bool appendType=true) 
            : base(position,orientation,joints, appendType ? "location." + type : type, name, uuid,parent,appendType) { }

            public new static Location FromDict(Dictionary<string, object> dct)
            {
                return new Location(
                    position: Position.FromDict((Dictionary<string, object>)dct["position"]),
                    orientation: Orientation.FromDict((Dictionary<string, object>)dct["orientation"]),
                    joints: (float[])dct["joints"],
                    type: (string)dct["type"],
                    name: (string)dct["name"],
                    uuid: (string)dct["uuid"],
                    appendType: false
                );
            }
        }

    }
}

File: evd_authoring_app/src/model/evdScript/data/location.js
import { Waypoint } from './waypoint';
import { Position, Orientation } from './geometry';
import Blockly from 'blockly';

export class Location extends Waypoint {

    /*
    * Data structure methods
    */

    static typeString() {
        return 'location.';
    }

    static fullTypeString() {
        return Waypoint.fullTypeString() + Location.typeString();
    }

    constructor(position=null, orientation=null, joints=null, type='', name='', 
                uuid=null, parent=null, appendType=true) 
    {
        super(
            position,
            orientation,
            joints,
            (appendType) ? 'location.'+type : type,
            name,
            uuid,
            parent,
            appendType
        );
    }

    static fromDict(dct) {
        return new Location(
            Position.fromDict(dct.position),
            Orientation.fromDict(dct.orientation),
            dct.joints,
            dct.type,
            dct.name,
            dct.uuid,
            null,
            false
        );
    }

    static BlocklyToolbox() {
        return { type: 'location' };
    }

    static BlocklyBlock(locations) {
        // TODO need to support type checking here
        // iterate through locations creating new fielddropdown entries
        // remove the unnamed option
        // look up syntax for blockly adding field options
        return { 
            name: 'Location',
            category: 'Location',
            block: {
                init: function () {
                    this.jsonInit({
                        type: "example_dropdown",
                        message0: "Location %1",
                        args0: [
                            {
                            type: "field_dropdown",
                            name: "FIELDNAME",
                            options: locations
                            }
                        ],
                        inputsInline: true,
                        colour: 260,
                        output: 'null',
                        tooltip: "location"
                    });
                },
            },
            generator: (block) => {
                var value_length = block.getFieldValue('name');
                var code = `console.log('Length is: ${value_length}')`;
                return [code, Blockly.JavaScript.ORDER_ATOMIC];
            }
        };
    }
}